File: converted_code/backend/src/main/java/com/zeriq/ai/controller/UserController.java
package com.zeriq.ai.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.zeriq.ai.model.User;
import com.zeriq.ai.service.UserService;

import java.util.Map;

@RestController
@RequestMapping("/api/users")
@CrossOrigin(origins = "*")
public class UserController {

    @Autowired
    private UserService userService;

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> payload) {
        String username = payload.get("username");
        String email = payload.get("email");
        
        if (username == null || email == null) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("error", "Username and email are required."));
        }
        
        User user = userService.loginOrCreate(username, email);
        return ResponseEntity.ok(user);
    }

    @GetMapping("/profile")
    public ResponseEntity<?> getProfile(@RequestHeader(value = "x-user-id", required = false) Long userId) {
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Unauthorized. Missing User ID header."));
        }
        
        return userService.getUserById(userId)
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("error", "User not found.")));
    }

    @PostMapping("/preferences")
    public ResponseEntity<?> updatePreferences(
            @RequestHeader(value = "x-user-id", required = false) Long userId,
            @RequestBody Map<String, Object> payload) {
        
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("error", "Unauthorized. Missing User ID header."));
        }

        try {
            Long courseId = Long.valueOf(payload.get("courseId").toString());
            String difficulty = payload.get("difficulty").toString();
            String preferredAiMode = payload.get("preferredAiMode").toString();
            
            User updatedUser = userService.updateCoursePreferences(userId, courseId, difficulty, preferredAiMode);
            return ResponseEntity.ok(updatedUser);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(Map.of("error", "courseId, difficulty, and preferredAiMode are required."));
        }
    }
}

File: src/backend/controllers/UserController.ts
import { Request, Response, Router } from "express";
import { UserService } from "../services/UserService";

export const userExtractor = (req: Request, res: Response, next: () => void) => {
  const userId = req.headers["x-user-id"];
  if (!userId) {
    res.status(401).json({ error: "Unauthorized. Missing User ID header." });
    return;
  }
  req.userId = Number(userId);
  next();
};

// Declaring property to extend Express Request object
declare global {
  namespace Express {
    interface Request {
      userId?: number;
    }
  }
}

export class UserController {
  private userService = new UserService();
  public router = Router();

  constructor() {
    this.router.post("/login", this.login.bind(this));
    this.router.get("/profile", userExtractor, this.getProfile.bind(this));
    this.router.post("/preferences", userExtractor, this.updatePreferences.bind(this));
  }

  private login(req: Request, res: Response): void {
    try {
      const { username, email } = req.body;
      if (!username || !email) {
        res.status(400).json({ error: "Username and email are required." });
        return;
      }
      const user = this.userService.loginOrCreate(username, email);
      res.json(user);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private getProfile(req: Request, res: Response): void {
    try {
      const user = this.userService.getUserById(req.userId!);
      if (!user) {
        res.status(404).json({ error: "User not found." });
        return;
      }
      res.json(user);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private updatePreferences(req: Request, res: Response): void {
    try {
      const { courseId, difficulty, preferredAiMode, preferredVoice } = req.body;
      if (!courseId || !difficulty || !preferredAiMode) {
        res.status(400).json({ error: "courseId, difficulty, and preferredAiMode are required." });
        return;
      }
      const updatedUser = this.userService.updateCoursePreferences(
        req.userId!,
        Number(courseId),
        difficulty,
        preferredAiMode,
        preferredVoice
      );
      res.json(updatedUser);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }
}
